src/explore-education-statistics-frontend/src/services/permalinkService.ts

```diff
@@ -0,0 +1,60 @@
+import { dataApi } from '@common/services/api';
+import { Dictionary } from '@common/types';
+import {
+  TableData,
+  GroupedFilterOptions,
+  TableDataQuery,
+} from '@common/services/tableBuilderService';
+import TimePeriod from '@common/services/types/TimePeriod';
+import {
+  Indicator,
+  LocationFilter,
+} from '@frontend/modules/table-tool/components/types/filters';
+import { TableHeadersFormValues } from '@frontend/modules/table-tool/components/TableHeadersForm';
+
+export interface FullTableMeta {
+  publicationName: string;
+  subjectId: string;
+  subjectName: string;
+  locations: LocationFilter[];
+  timePeriodRange: TimePeriod[];
+  filters: Dictionary<{
+    legend: string;
+    hint?: string;
+    options: GroupedFilterOptions;
+    totalValue?: string;
+  }>;
+  indicators: Indicator[];
+  footnotes: {
+    id: number;
+    label: string;
+  }[];
+}
+
+export interface PermalinkCreate extends TableDataQuery {
+  configurations: {
+    tableHeadersConfig: TableHeadersFormValues;
+  };
+}
+
+export interface FullTable /* ™ */ {
+  title?: string;
+  subjectMeta: FullTableMeta;
+  results: TableData['results'];
+}
+
+export interface Permalink {
+  id: string;
+  title: string;
+  created: string;
+  fullTable: FullTable;
+  configuration: {
+    tableHeadersConfig?: TableHeadersFormValues;
+  };
+}
+
+export default {
+  getPermalink(publicationSlug: string): Promise<Permalink> {
+    return dataApi.get(`Permalink/${publicationSlug}`);
```

Comment:
The returned `Permalink` type is lying to us here as it hasn't actually been mapped correctly at this point. Rather than map the correct types later on within the components, we should just construct them all here.

This is particularly useful for avoiding fragmentation and wastefully re-constructing the same objects in multiple places.

Reply:
Is there a Typescript way to do this?
Do I need to create a class that will construct the response into a full Permalink?

Reply:
No there isn't a Typescript-specific way of doing this (it's just Javascript with types after all).

You'll need to map over everything and then construct the filter objects in the same way that you have been doing across the various components, but just in here instead.